### PR TITLE
Ticket1567 create meaningful script 

### PR DESCRIPTION
**Pull request by ThomasLohnert**
### Description of work

- Implemented functionality to create meaningful scripts from script settings & row parameters.
- Added unit tests to check script correctness with various settings.
- Added some helpful constraints & auto fill values to GUI: disc sample geometry links width & height; adding a row (via setting sample position) fills columns with default values taken from default values in method declaration; deleting a row (via clearing sample position) clears values from rest of columns.

NOTE: the python method calls in the generated script anticipate some slight changes to the dosans/dotrans methods on the instruments (based off current code on LARMOR). Specifically, the "uamps" parameter needs to be renamed to the more accurate "waitfor" and a "change_period" parameter needs to be added (to be used in the future).

### To test

https://github.com/ISISComputingGroup/IBEX/issues/1567

### Acceptance criteria

Automatically generated script runs data collection with the right settings in the right order.
Original Script Generator for comparison: [scriptgen.zip](https://github.com/ISISComputingGroup/ibex_gui/files/663540/scriptgen.zip)
(note: does not implement "loop over each run" setting for genie_python scripts)

I also created a graphic illustrating the expected running order for different settings on a table with 2 rows and 2 loops per sample. You may refer to this to check correctness:

S(A)=SANS(sample A), T(B)=TRANS(sample B) and so on...
script_running_order

---

#### Code Review

- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/GUI-Coding-Conventions)? Is it well structured with small focussed classes/methods/functions?
- [x] Are there unit tests in place? Are the unit tests small and test the a class in isolation?
- [x] Are there [system tests](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/System-Testing-with-RCPTT) in place? Do they test a minimal set of functionality and leave the gui as close as possible to its original state?
- [x] Did any existing system test break as a result of the current changes? 
- [x] Have the changes been documented in the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev). If so, do they describe the changes appropriately?
- [x] Has the [manual system tests spreadsheet](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Manual-system-tests) been updated?

### Functional Tests

- [x] Do changes function as described? Add comments below that describe the tests performed.
- [x] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has developer documentation been updated if required?

### Final Steps
- [ ] Reviewer has moved the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev) entry for this ticket in the "Changes merged into master" section

### CODE REVIEW

File: base/uk.ac.stfc.isis.ibex.scriptgenerator/src/uk/ac/stfc/isis/ibex/scriptgenerator/PythonBuilder.java [hunk 1]
```diff
@@ -54,26 +270,67 @@ public void setRows(Collection<Row> rows) {
...
+		this.settings = settings;
+	}
+	
 	/**
 	 * Returns the script back to the view to be displayed.
 	 * @return the completed script
 	 */
 	public String getScript() {
-        return createScript();
+		return script.toString();
```
Comment (AdrianPotter) [inline base/uk.ac.stfc.isis.ibex.scriptgenerator/src/uk/ac/stfc/isis/ibex/scriptgenerator/PythonBuilder.java]: It looks like getScript is unused now. Can it be removed entirely?

File: base/uk.ac.stfc.isis.ibex.scriptgenerator/src/uk/ac/stfc/isis/ibex/scriptgenerator/PythonBuilder.java [hunk 2]
```diff
@@ -54,26 +270,67 @@ public void setRows(Collection<Row> rows) {
...
+     * Calls internal build script methods and adds these returned strings to
+     * the main script.
      * 
-     * Currently does not create a valid script!
+     * @return The Script.
      */
-    private String createScript() {
-        StringBuilder sb = new StringBuilder();
+    public String createScript() {
+		script = new StringBuilder();
```
Comment (AdrianPotter) [inline base/uk.ac.stfc.isis.ibex.scriptgenerator/src/uk/ac/stfc/isis/ibex/scriptgenerator/PythonBuilder.java]: Assuming getScript() can be removed, script no longer needs to be a class variable and can be made local

File: base/uk.ac.stfc.isis.ibex.scriptgenerator/src/uk/ac/stfc/isis/ibex/scriptgenerator/PythonBuilder.java [hunk 3]
```diff
@@ -54,26 +270,67 @@ public void setRows(Collection<Row> rows) {
...
+		
+		script.append(generateHeader());
 		
-		for (Row row : rows) {
-            sb.append("position = " + row.getPosition() + System.lineSeparator());
-		}
+        script.append(indent(buildSamplePar("height", settings.getSampleHeight().toString())));
+        script.append(indent(buildSamplePar("width", settings.getSampleWidth().toString())));
+        script.append(indent(buildSamplePar("geometry", settings.getGeometry().toString())));
```
Comment (AdrianPotter) [inline base/uk.ac.stfc.isis.ibex.scriptgenerator/src/uk/ac/stfc/isis/ibex/scriptgenerator/PythonBuilder.java]: You could avoid having to do all of these toStrings if `buildSamplePar` did it instead

File: base/uk.ac.stfc.isis.ibex.scriptgenerator/src/uk/ac/stfc/isis/ibex/scriptgenerator/PythonBuilder.java [hunk 4]
```diff
@@ -54,26 +270,67 @@ public void setRows(Collection<Row> rows) {
...
+                script.append(indent(buildSequential(true)));
+                break;
+            case TRANS:
+                script.append(indent(buildSequential(false)));
+                break;
+            case ALTSANS:
+                script.append(indent(buildAlternating(true)));
+                break;
+            case ALTTRANS:
+                script.append(indent(buildAlternating(false)));
```
Comment (AdrianPotter) [inline base/uk.ac.stfc.isis.ibex.scriptgenerator/src/uk/ac/stfc/isis/ibex/scriptgenerator/PythonBuilder.java]: If the logic of adding something to the script changes:
    - Script to longer a list
    - Indent method interface changes
    - Want to apply additional methods when adding something to the script
then you'll have to change all of these. It'd be better if it were handled with a small private method

File: base/uk.ac.stfc.isis.ibex.scriptgenerator/src/uk/ac/stfc/isis/ibex/scriptgenerator/PythonBuilder.java (hunk 3 above)
Comment (AdrianPotter) [inline base/uk.ac.stfc.isis.ibex.scriptgenerator/src/uk/ac/stfc/isis/ibex/scriptgenerator/PythonBuilder.java]: See also comment below about `script.append.indent(...` pattern

**Comment by AdrianPotter** [comment]
For the overall design, the `PythonBuilder` for me is overly long and doesn't fulfil single responsibility. A clearer design would be:

`PythonBuilder 1->1 Script  1->* ScriptRow`

**Comment by AdrianPotter** [comment]
As part of the rework, please could you provide instructions/location for the original script generator so that the output can be verified as being unchanged.

**Comment by DominicOram** [comment]
Looks good. 
Checkstyle increase is due to test files, which is fine.
Can you add some bits to the dev wiki? At least a link to Matt's original program and the diagram above.

**Pull request merged**